File: semester 4/Algorithms_and_data_structures/c++/shop(8).cpp
/**
 * https://contest.yandex.ru/contest/35179/problems/H/
 */
#include <iostream>

using namespace std;

void swap(int *array, int firstIndex, int secondIndex) {
    int temp = array[firstIndex];
    array[firstIndex] = array[secondIndex];
    array[secondIndex] = temp;
}

void quickSort(int *array, int lowIndex, int highIndex) {
    if (lowIndex >= highIndex) return;

    const int pivot = array[highIndex];
    int leftPointer = lowIndex;
    int rightPointer = highIndex - 1;

    while (leftPointer < rightPointer) {
//while (rawArray[leftPointer] <= pivot);
        while (array[leftPointer] <= pivot && leftPointer < rightPointer)
            leftPointer++;
//while (rawArray[rightPointer] >= pivot)
        while (array[leftPointer] >= pivot && leftPointer < rightPointer)
            rightPointer--;

        swap(array, leftPointer, rightPointer);
    }


    if (array[leftPointer] > array[highIndex])
        swap(array, leftPointer, highIndex);
    else
        leftPointer = highIndex;

    quickSort(array, lowIndex, leftPointer - 1);
    quickSort(array, leftPointer + 1, highIndex);
}

int compare(const void *a, const void *b) {
    return (*(int *) a - *(int *) b);
}

int main8() {
    int productCount, saleRate, saleCount = 0;
    cin >> productCount;
    cin >> saleRate;
    int productsCost[productCount];

    for (int i = 0; i < productCount; i++) {
        cin >> productsCost[i];
    }

   // quickSort(productsCost, 0, productCount - 1);
    qsort(productsCost, productCount, sizeof(int), compare);

    long amountToPay = 0;

    for (int i = productCount - 1; i >= 0; i--) {
        saleCount++;
        if (saleCount % saleRate == 0) continue;
        amountToPay += productsCost[i];
    }

    cout << amountToPay;
}

File: semester 4/Algorithms_and_data_structures/kotlin/Shop(8).kt
/**
 * https://contest.yandex.ru/contest/35179/problems/H/
 */

package agronom

import java.util.*

fun main() {
    val scanner = Scanner(System.`in`)
    val productCount = 5
    val saleRate = 2
    val test1: Array<Int> = listOf(200, 100, 1000, 400, 100).toTypedArray()

    quickSort(test1, 0, test1.size - 1)

    var amountToPay: Long = 0
    var saleCount = 0

    for (i in test1.indices){
        saleCount++
        if (saleCount % saleRate == 0) continue
        amountToPay += test1[i]
    }

    println(amountToPay)

}

private fun swap(rawArray: Array<Int>, leftIndex: Int, rightIndex: Int) {
    val temp = rawArray[leftIndex]
    rawArray[leftIndex] = rawArray[rightIndex]
    rawArray[rightIndex] = temp
}


fun quickSort(array: Array<Int>, lowIndex: Int, highIndex: Int) {
    if (lowIndex >= highIndex) return

    val pivot = array[highIndex]
    var leftPointer = lowIndex
    var rightPointer = highIndex - 1

    while (leftPointer < rightPointer) {
        //while (rawArray[leftPointer] <= pivot)
        while (array[leftPointer] <= pivot && leftPointer < rightPointer)
            leftPointer++
        //while (rawArray[rightPointer] >= pivot)
        while (array[leftPointer] >= pivot && leftPointer < rightPointer)
            rightPointer--

        swap(array, leftPointer, rightPointer)
    }


    if (array[leftPointer] > array[highIndex])
        swap(array, leftPointer, highIndex)
    else
        leftPointer = highIndex

    quickSort(array, lowIndex, leftPointer - 1)
    quickSort(array, leftPointer + 1, highIndex)
}